File: services/16-echo/internal/auth/middleware.go
package auth

import (
	"net/http"
	"strings"

	"github.com/labstack/echo/v4"
	"github.com/yarova-ca/16-echo/internal/db"
)

// RequireAuth is an Echo middleware (func(echo.HandlerFunc) echo.HandlerFunc).
//
// Authentication order:
//  1. Authorization: Bearer <JWT> header — verified via VerifyToken.
//  2. X-API-Key header — looked up in the users table.
//
// On success: sets c.Set("user", *Claims) and calls next.
// On failure: returns 401 JSON response.
func RequireAuth(next echo.HandlerFunc) echo.HandlerFunc {
	return func(c echo.Context) error {
		// --- Attempt 1: Bearer JWT ---
		if authHeader := c.Request().Header.Get("Authorization"); authHeader != "" {
			parts := strings.SplitN(authHeader, " ", 2)
			if len(parts) == 2 && strings.EqualFold(parts[0], "bearer") {
				claims, err := VerifyToken(parts[1])
				if err != nil {
					return c.JSON(http.StatusUnauthorized, map[string]string{"error": "invalid token"})
				}
				c.Set("user", claims)
				return next(c)
			}
		}

		// --- Attempt 2: X-API-Key header ---
		if apiKey := c.Request().Header.Get("X-API-Key"); apiKey != "" {
			database := db.GetDB()
			var user db.User
			result := database.Where("api_key = ?", apiKey).First(&user)
			if result.Error != nil {
				return c.JSON(http.StatusUnauthorized, map[string]string{"error": "invalid api key"})
			}
			claims := &Claims{UserID: user.ID, Email: user.Email, Name: user.Name}
			c.Set("user", claims)
			return next(c)
		}

		return c.JSON(http.StatusUnauthorized, map[string]string{"error": "authentication required"})
	}
}

// GetClaims extracts the authenticated Claims from the Echo context.
func GetClaims(c echo.Context) *Claims {
	val := c.Get("user")
	if val == nil {
		return nil
	}
	claims, _ := val.(*Claims)
	return claims
}

File: services/28-rust-grpc/src/auth/middleware.rs
use tonic::{Request, Status};
use sqlx::PgPool;

use crate::auth::jwt::verify_token;

/// Authenticated user extracted from gRPC metadata.
#[derive(Debug, Clone)]
pub struct AuthUser {
    pub id: String,
    pub email: String,
    pub name: String,
}

/// Tonic interceptor function — validates JWT or API key from gRPC metadata.
///
/// gRPC metadata is the equivalent of HTTP headers.
/// Clients send: metadata.insert("authorization", "Bearer <token>".parse()?)
///               metadata.insert("x-api-key", "<key>".parse()?)
///
/// On valid JWT: attaches AuthUser to request extensions, returns Ok.
/// On valid API key: DB lookup, attaches user, returns Ok.
/// On missing or invalid credential: returns Err(Status::unauthenticated).
///
/// Usage: Server::builder().layer(tonic::service::interceptor(auth_interceptor))
pub fn auth_interceptor(mut req: Request<()>) -> Result<Request<()>, Status> {
    let metadata = req.metadata();

    // --- Attempt 1: Bearer JWT ---
    if let Some(auth_value) = metadata.get("authorization") {
        let auth_str = auth_value
            .to_str()
            .map_err(|_| Status::unauthenticated("Invalid authorization header encoding"))?;

        if let Some(token) = auth_str.strip_prefix("Bearer ") {
            let claims = verify_token(token)
                .map_err(|_| Status::unauthenticated("Invalid or expired JWT token"))?;
            req.extensions_mut().insert(AuthUser {
                id: claims.sub,
                email: claims.email,
                name: claims.name,
            });
            return Ok(req);
        }
    }

    // --- Attempt 2: X-API-Key ---
    // Note: API key lookup requires async DB access.
    // For async gRPC services, perform this check inside the handler using
    // the PgPool from tonic State, or use a tower middleware layer instead.
    if let Some(api_key_value) = metadata.get("x-api-key") {
        if api_key_value.to_str().is_ok() {
            // Mark request as needing async API key validation in the handler.
            // The handler calls validate_api_key(pool, key).await.
            return Ok(req);
        }
    }

    Err(Status::unauthenticated(
        "Authentication required. Provide Bearer token or x-api-key metadata.",
    ))
}

/// Validates an API key against the database.
/// Call this inside an async gRPC handler when x-api-key metadata is present.
pub async fn validate_api_key(pool: &PgPool, api_key: &str) -> Result<AuthUser, Status> {
    let row = sqlx::query_as::<_, (String, String, String)>(
        "SELECT id, email, name FROM users WHERE api_key = $1",
    )
    .bind(api_key)
    .fetch_optional(pool)
    .await
    .map_err(|_| Status::internal("Auth check failed"))?;

    match row {
        Some((id, email, name)) => Ok(AuthUser { id, email, name }),
        None => Err(Status::unauthenticated("Invalid API key")),
    }
}
